### PR TITLE
feat(core+cli): ADR-091 pack pending-verification install→lint promotion (closes #1684)

### LINKED ISSUE #1684: core: ADR-091 pack pending-verification install→lint promotion flow
﻿## Context

The cloud-compile bootstrap mode for mmnto-ai/totem#1682 (Stage 4 verifier). Per [ADR-091 ┬º"Bootstrap Semantics" mode 2](https://github.com/mmnto-ai/totem-strategy/blob/main/adr/adr-091-ingestion-pipeline-refinements.md#bootstrap-semantics): when packs are compiled in the cloud, the consumer's codebase is not reachable at compile time. New rules enter `pending-verification` state when installed; the next `totem lint` or `totem review` run on the consumer's machine acts as the codebase verifier.

The pack's attestation (per ADR-089) is not bypassed; only the pack author's local verification is deferred to the installer's first lint. Together with ADR-089's zero-trust default and Sigstore signing, this closes the loop: installed pack rules cannot fire on consumer code without having been verified against it.

### Install-time stamping

- [ ] Pack rules installed via `totem install pack/<name>` are stamped `pending-verification` in the consumer's `compiled-rules.json` (or equivalent local manifest entry). Stamp format: a status enum value, not a flag ΓÇö composes cleanly with the existing `active` / `archived` / `untested-against-codebase` set from mmnto-ai/totem#1682.
- [ ] `pending-verification` rules do NOT fire during `totem lint` or `totem review` until promoted. They remain inert from the consumer's perspective; the install message explicitly states this.
- [ ] Install output prints: "Run `totem lint` to activate pack rules" message (per ADR-091 ┬º"Risk Assessment" mitigation for stale-pending).

### First-lint promotion path

- [ ] First `totem lint` or `totem review` after install detects `pending-verification` rules and runs the mmnto-ai/totem#1682 verifier against each.
- [ ] Stage 4's four outcomes apply identically (per mmnto-ai/totem#1682 acceptance criteria):
  - No matches ΓåÆ promoted to `active` with `confidence = 'untested-against-codebase'`.
  - Out-of-scope baseline matches ΓåÆ demoted to `archived` with `reasonCode = 'stage4-out-of-scope-match'`. Rule disabled locally; the pack itself is not modified upstream.
  - In-scope `badExample`-shape matches ΓåÆ promoted to `active` with `confidence = 'high'`.
  - In-scope non-`badExample` matches ΓåÆ promoted to `active` as Candidate Debt with forced `severity: warning`.
- [ ] Verification results written to `.totem/verification-outcomes.json` with an entry per pack-installed rule. Schema includes: rule hash, verification timestamp, outcome, baseline-match paths if any.
- [ ] Subsequent `totem lint` runs read `.totem/verification-outcomes.json` and skip re-verification for rules with a recorded outcome. Re-verification triggers: explicit `totem rule promote --re-verify <hash>` or rule-content change (hash change).

### CI-first installations

- [ ] When the first lint runs in CI (rather than locally), verification still triggers and writes `.totem/verification-outcomes.json`. The file is committable so subsequent CI runs and local runs share the verification result. Document this in `docs/wiki/cli-reference.md` install-section additions.

### Consumer-facing observability

- [ ] `totem doctor` surfaces are covered in the companion mmnto-ai/totem#1685 ticket; this ticket only delivers the installΓåÆlint state machine and the `.totem/verification-outcomes.json` schema.

## Dependencies

- **Hard depends on mmnto-ai/totem#1682** (verifier must exist before installΓåÆlint can promote).
- **Hard depends on mmnto-ai/totem#1683** (baseline semantics must be defined for the verifier outcomes to apply consistently to pack rules).
- **Soft depends on mmnto-ai/totem#1492** (Sigstore signing). Not a blocker ΓÇö `pending-verification` is independent of attestation. But the user-facing message in install output should reflect both gates once #1492 ships.
- **Blocks mmnto-ai/totem#1685** (UX surfaces consume `pending-verification` state defined here).

## Out of Scope

- Stage 4 verifier itself ΓÇö see mmnto-ai/totem#1682.
- Baseline schema and config ΓÇö see mmnto-ai/totem#1683.
- `totem doctor` reporting and 30-day stale warning ΓÇö see mmnto-ai/totem#1685.
- Pack signing / attestation flow ΓÇö already in scope on mmnto-ai/totem#1492.

## References

- ADR-091 ┬º"Bootstrap Semantics" ΓÇö `mmnto-ai/totem-strategy/blob/main/adr/adr-091-ingestion-pipeline-refinements.md`
- ADR-089 Zero-Trust Agent Governance (pack installer flow context) ΓÇö `mmnto-ai/totem-strategy/blob/main/adr/adr-089-zero-trust-agent-governance.md`
- Decomposition audit doc ΓÇö `mmnto-ai/totem-strategy/blob/main/audits/internal/2026-04-25-adr-091-decomposition.md`
- mmnto-ai/totem#1492 (pack Sigstore signing ΓÇö orthogonal but adjacent)
- mmnto-ai/totem#1609 (`packages/pack-agent-security` private:false flip ΓÇö gated on #1492)

### PR DESCRIPTION
**Pull request by satur8d**
## Summary

Closes the cloud-compile bootstrap gap that ADR-091 § Bootstrap Semantics defined. Pack rules installed via `totem install` cannot be trusted to fire on the consumer's codebase until Stage 4 verifies them locally — so they now enter the manifest as `'pending-verification'` and the next `totem lint` runs the verifier and promotes them per outcome.

Six commits split along the spec's task plan (T1–T6 in `.totem/specs/1684.md`):

- **T1** `020f1291` — `CompiledRule.status` enum extended with `'pending-verification'` (alongside `'active' | 'archived' | 'untested-against-codebase'`); `verification-outcomes.ts` defines Zod schemas for the new committable side-table with a compile-time alignment guard against `Stage4Outcome` drift.
- **T2** `4e7f75dc` — `readVerificationOutcomes` / `writeVerificationOutcomes` persistence with self-healing on corrupt-read (Invariant #10), atomic temp+rename writes, and recursive object-key sort for byte-stable serialization (Invariant #11).
- **T3** `f319147d` — `installCommand` force-stamps every pack rule `'pending-verification'` regardless of the status the pack shipped with, and prints `Run \`totem lint\` to activate pack rules` after the install-success line (Invariant #8 exact-match locked via constant + test).
- **T4** `8e0fd6e1` — `loadCompiledRules` filter (lint-execution path) extended to also drop `'pending-verification'` rules; `loadCompiledRulesFile` (admin path) keeps them visible so the interceptor can find them.
- **T5** `08dd8823` — `first-lint-promote.ts` exports `promotePendingRules(rules, deps)` with per-rule verifier-throw isolation (Invariant #7), empty-pending fast path (Invariant #9), and memoization across runs (Invariant #4) keyed on `lessonHash` (Invariant #5 hash-invalidation re-verify). CLI runner at `first-lint-promote-runner.ts` builds Stage 4 deps and wires into `lintCommand` before `getDiffForReview`.
- **T6** `6159046c` — Changeset (minor bump on the fixed group of `@mmnto/{totem,cli,mcp,pack-rust-architecture,pack-agent-security}`), wiki entry for `totem install`, and monorepo path correctness (resolve `repoRoot` eagerly so `.totemignore` and the manifest-exclusion entry use repo-relative paths instead of mis-joining against `configRoot` in monorepo subpackages — addresses two findings from the first `totem review` pass).

## Naming-collision context

The original ADR-091 draft specified `.totem/rule-metrics.json`, but `packages/core/src/rule-metrics.ts` already exists as a per-machine telemetry-cache module (`triggerCount`, `suppressCount`, `evaluationCount`) with a gitignored `.totem/cache/rule-metrics.json` lifetime. Multi-agent decision (option B): rename the new committable verification state to `.totem/verification-outcomes.json` + new `verification-outcomes.ts` module. ADR-091 § 65 was amended in `mmnto-ai/totem-strategy` main to match.

## Test plan

- [x] **`@mmnto/totem`**: 1753/1753 tests pass (was 1719; +34 new — schema validation, persistence round-trips, byte-stable serialization, four Stage4Outcome mappings, memoization-skip, hash-invalidation re-verify, verifier-throw isolation, empty-pending fast path, filter regression).
- [x] **`@mmnto/cli`**: 1962/1962 tests pass (was 1958; +4 new — stamper purity, activation-message exact-substring lock-in, install-flow regression).
- [x] `totem lint` PASS (0 errors, 3 warnings — substrate-rule false positives in the spec doc + test code, tier-3 follow-up).
- [x] `totem review` PASS via `gemini-3.1-pro-preview` against the full branch diff.

## Known follow-ups

1. Substrate-rule false positives on internal scaffolding (`os.tmpdir()` / `fs.readFileSync` in tests; "absolute claims" rule firing on RFC-2119 prose in `.totem/specs/`) — file as tier-3.
2. The same monorepo-subpackage manifest-exclusion path issue exists in `compile.ts:663`. T6 fixed it on my code; mirror to `compile.ts` as tier-3.
3. `totem doctor` Stage 4 UX surfaces (`mmnto-ai/totem#1685`) and Stage 4 perf hardening (`mmnto-ai/totem#1686`) remain — tracked in their own tickets.

Closes #1684.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by satur8d** [comment]
@coderabbitai full review
/gemini review

ADR-091 § Bootstrap Semantics — pack pending-verification install→lint promotion (closes #1684). Six commits (T1–T6 from `.totem/specs/1684.md`), three new core modules (`verification-outcomes.ts` schemas + persistence; `first-lint-promote.ts` interceptor) plus CLI runner and lint-command wiring. Full diff is the install-time stamping → first-lint Stage 4 sweep → outcomes-memoized subsequent runs path; +38 tests across core + cli; local `totem lint` 0 errors and `totem review` PASS via Gemini 3.1 Pro.

File: packages/cli/src/commands/first-lint-promote-runner.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  const {
+    loadCompiledRulesFile,
+    parseStage4BaselineDirectives,
+    promotePendingRules,
+    resolveStage4Baseline,
+    safeExec,
+    saveCompiledRulesFile,
+    STAGE4_MANIFEST_EXCLUSIONS,
+    verifyAgainstCodebase,
+  } = await import('@mmnto/totem');
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/cli/src/commands/first-lint-promote-runner.ts:41]: Fixed in `f0373a97`. Imported `sanitizeForTerminal` from `@mmnto/totem` and wrapped the `onWarn` callback so promotion-warning text passes through the strip before reaching `log.warn`.

File: packages/core/src/first-lint-promote.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+  // Empty-pending fast path (Invariant #9): skip the outcomes-file read
+  // when there's nothing to verify. Returns the original array unchanged
+  // — callers can shallow-equal compare to detect no-op.
+  if (pendingIndices.length === 0) {
+    return { mutatedRules: [...rules], ...NOOP_RESULT_TEMPLATE };
+  }
+
+  const onWarn = deps.onWarn ?? (() => {});
+  const now = deps.now ?? (() => new Date());
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/first-lint-promote.ts:116]: Fixed by updating the JSDoc rather than the code (`f0373a97`). The doc claimed `console.warn` as the default but the project lint rule prohibits `console.*` calls in `@mmnto/totem` core, so the noop default is correct — the doc was wrong. JSDoc now states "defaults to a no-op when omitted; pass an explicit logger to surface diagnostics." The CLI runner already wires `log.warn` (post-sanitize).

File: packages/core/src/first-lint-promote.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+      // never written back to the manifest (interrupted prior pass) or the
+      // rule was re-stamped pending by a reinstall — either way, the
+      // recorded outcome is still authoritative.
+      outcomeEntry = memoized;
+    } else {
+      // No memoized outcome (or hash mismatch on a stale entry). Invoke the
+      // verifier. Per-rule try/catch isolates failures (Invariant #7).
+      try {
+        const result = await deps.verifier(rule);
+        verifierInvocations += 1;
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/first-lint-promote.ts:143]: Fixed in `f0373a97`. Moved `verifierInvocations += 1` before the `await deps.verifier(rule)` so throws are counted as work attempted. JSDoc updated: "Includes calls that threw, since the metric reports work attempted rather than work succeeded — pair with `verifierFailures` to derive the success count." Test `first-lint-promote.test.ts:269` updated to assert the new contract.

File: packages/cli/src/commands/first-lint-promote-runner.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  // compiled-rules file (false-positive matches against rules' own
+  // `badExample` text).
+  const repoRoot: string = safeExec('git', ['rev-parse', '--show-toplevel'], {
+    cwd,
+    env: { ...process.env, LC_ALL: 'C' },
+  });
+
+  const activeManifestPath = path.relative(repoRoot, manifestPath).replace(/\\/g, '/');
+  const manifestExclusionSet = new Set<string>([...STAGE4_MANIFEST_EXCLUSIONS, activeManifestPath]);
+
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/cli/src/commands/first-lint-promote-runner.ts:106]: Fixed in `f0373a97`. Added `activeOutcomesPath` to the manifest-exclusion set alongside `activeManifestPath`. Now committed `.totem/verification-outcomes.json` is filtered before Stage 4 sees the file list.

File: packages/core/src/verification-outcomes.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  const result = VerificationOutcomesFileSchema.safeParse(parsed);
+  if (!result.success) {
+    const issues = result.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join('; ');
+    onWarn?.(
+      `Schema validation failed for ${filePath}: ${issues}. Treating as empty; next write will overwrite.`,
+    );
+    return {};
+  }
+  return result.data.outcomes;
+}
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/verification-outcomes.ts:156]: Fixed in `f0373a97`. After schema-validation, the loader now iterates `result.data.outcomes` and drops any entry whose `ruleHash` does not match its file-level key, with a warn through `onWarn`. New test `verification-outcomes.test.ts` covers the tampered-key scenario.

File: packages/core/src/verification-outcomes.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+): void {
+  const dir = path.dirname(filePath);
+  fs.mkdirSync(dir, { recursive: true });
+
+  const file: VerificationOutcomesFile = { version: 1, outcomes };
+  const json = `${JSON.stringify(canonicalizeForSerialization(file), null, 2)}\n`;
+
+  const tmpPath = `${filePath}.tmp`;
+  fs.writeFileSync(tmpPath, json, 'utf-8');
+  fs.renameSync(tmpPath, filePath);
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/verification-outcomes.ts]: Fixed in `f0373a97`. Temp filename now uses `${filePath}.${process.pid}.${Date.now().toString(36)}.tmp` and the writeSync/renameSync pair is wrapped in try/finally with `fs.rmSync(tmpPath, { force: true })` cleanup. Concurrent CI lint matrix builds can no longer clobber each other's temp file.

**Comment by satur8d** [comment]
@gemini-code-assist Thank you for the review. Addressing all three findings on commit `005b739c` (preceded by `f0373a97` which carries the bulk of the changes):

**1. `verification-outcomes.ts:161` — DRY against `canonicalStringify`.** Valid signal, suggested-fix didn't compile (existing `canonicalStringify` was 1-arg, not 2-arg). Resolved by extracting a shared `canonicalizeKeys` primitive in `compile-manifest.ts` and giving `canonicalStringify` an optional `indent` parameter that routes through `JSON.stringify(canonicalizeKeys(value), null, indent)`. The minified hash-payload path is unchanged; `writeVerificationOutcomes` now calls `canonicalStringify(file, 2)`. Both helpers are exported from `@mmnto/totem`. Six new unit tests cover the recursion, indent round-trip, and primitive/undefined-property edge cases (`compile-manifest.test.ts`).

**3. `first-lint-promote-runner.ts:78` — unprotected `safeExec` for git root.** Fixed per the cited styleguide rule. Wrapped the call in try/catch; on failure we now `log.error('Totem Error', ...)` and set `process.exitCode = 1` before returning, rather than letting the exception propagate. Trailing `// totem-context: intentional cleanup — CLI rejects throw at top-level per .gemini/styleguide.md` directive on the `safeExec` line documents the deliberate non-rethrow.

**7. `first-lint-promote-runner.ts:136` — atomic write for `compiled-rules.json` (deferred).** The fix is correct in principle but in the wrong place. `saveCompiledRulesFile` is the shared helper used by every writer of `.totem/compiled-rules.json` repo-wide (see `compiler.ts:210`); wrapping atomicity at this single call site leaves every other caller exposed. The architecturally honest fix is to centralize atomic-rename inside `saveCompiledRulesFile` itself, mirroring the temp+rename pattern that PR #1787 introduced for `writeVerificationOutcomes`. Filed as tier-3 follow-up `#1788` — out of scope for this PR (which is the install→lint promotion path) but the right next step.

All fixes verified locally: 1760/1760 core tests + 1962/1962 cli tests pass; `totem lint` PASS (0 errors); `totem review` PASS (no critical findings).

**Comment by satur8d** [comment]
@coderabbitai full review

/gemini review
